src/Select.purs

```diff
@@ -90,266 +78,317 @@ derive instance ordVisibility :: Ord Visibility
+  (UseGet SelectState
+  (UseState SelectState hooks)))
+
+derive instance newtypeUseSelect :: Newtype (UseSelect hooks) _
+
+-- | A `SelectInput` value whose defaults can be overrided. **Note**:
+-- | `getItemCount` must be overrided:
+-- |
+-- | Default values are:
+-- | ```
+-- | { inputType: Toggle
+-- | , search: Nothing
+-- | , debounceTime: Nothing
+-- | , getItemCount: pure 0 -- this must be overrided!
+-- | , pushNewSearch: \_ -> pure unit
+-- | , pushVisibilityChanged: \_ -> pure unit
+-- | , pushSelectedIdxChanged: \_ -> pure unit
+-- | }
+-- | ```
+-- |
+-- | Example:
+-- | ```
+-- | events <- useEvent
+-- | select <- useSelect $ selectInput
+-- |   { getItemCount = pure (length items)
+-- |   , pushNewSearch = events.push
+-- |   }
+-- | ```
+selectInput :: forall m. SelectInput m
+selectInput = do
+  let
+    ignoreEvent :: forall a. a -> HookM m Unit
+    ignoreEvent = \_ -> pure unit
+
+  { inputType: Toggle
+  , search: Nothing
+  , debounceTime: Nothing
+  , getItemCount: pure 0
+  , pushNewSearch: ignoreEvent
+  , pushVisibilityChanged: ignoreEvent
+  , pushSelectedIdxChanged: ignoreEvent
   }
 
-component
-  :: forall st query action slots input msg m
+useSelect
+  :: forall m toggleProps itemProps containerProps inputProps
    . MonadAff m
-  => Row.Lacks "debounceRef" st
-  => Row.Lacks "visibility" st
-  => Row.Lacks "highlightedIndex" st
-  => (input -> Input st)
-  -> Spec st query action slots input msg m
-  -> H.Component HH.HTML (Query query slots) input msg m
-component mkInput spec = H.mkComponent
-  { initialState: initialState <<< mkInput
-  , render: spec.render
-  , eval: H.mkEval
-      { handleQuery: handleQuery spec.handleQuery
-      , handleAction: handleAction spec.handleAction spec.handleEvent
-      , initialize: Just (Initialize spec.initialize)
-      , receive: map Action <<< spec.receive
-      , finalize: map Action spec.finalize
+  => SelectInput m
+  -> Hook m UseSelect
+        (SelectReturn m toggleProps itemProps containerProps inputProps)
+useSelect inputRec =
+  let
+    initialSearchValue = fromMaybe "" inputRec.search
+    debounceTime = fromMaybe mempty inputRec.debounceTime
+  in Hooks.wrap Hooks.do
+    state /\ modifyState <- useState
+      { search: initialSearchValue
+      , visibility: Off
+      , highlightedIndex: Nothing
+      }
+    getState <- useGet state
+
+    searchDebouncer <- useDebouncer debounceTime \lastSearchState -> Hooks.do
+      case inputRec.inputType of
+        Text -> do
+          -- modifyState (_ { highlightedIndex = (Just 0) })
+          inputRec.pushNewSearch lastSearchState
+
+        -- Key stream is not yet implemented. However, this should capture user
+        -- key events and expire their search after a set number of milliseconds.
+        _ -> pure unit
+
+    Hooks.pure
+      -- state
+      { search: state.search
+      , visibility: state.visibility
+      , highlightedIndex: state.highlightedIndex
+
+      -- actions
```

Comment:
Since there are only three actions being returned to the user, and the others are being used to set props, it seems like some refactoring could be done to reduce all the argument passing happening here.

Are you open to moving functions that depend on `state` up into let bindings here, to skip the argument passing, while leaving others which don't depend on it directly (like `setFocus`) in the where clause?

As a rule of thumb, anything that would take `state` or `modifyState` as an argument can be moved up into a let binding, and everything else can remain in the  `where` clause.

That way, once you've defined the actions that depend on state in `let`s you can just pass the resulting `HookM m Unit` functions directly to things like the `toggleProps` function, which reduces all the coordination being done among these various functions.

Still others can just be defined inline -- for example, `clearSearch` might as well just be defined inline as `clearSearch: modifyState _ { search = "" }` 

So, for example:

```purs
let
  setVisibility :: Visibility -> HookM m Unit
  setVisibility v = do
    when (state.visibility /= v) do
      modifyState (_ { visibility = v, highlightedIndex = Just 0 })
      inputRec.pushVisibilityChanged v

Hooks.pure
  { setVisibility
  , clearSearch: modifyState _ { search = "" }
  , toggleProps: toggleProps setVisibility toggleClick key
  }
where
toggleProps :: (Visibility -> HookM m Unit) -> ...
```

Reply:
I've inlined `clearSearch` in latest commits.

Reply:
I haven't done the other work yet because when I was thinking about this one, I [wondered whether `useGet` is necessary in case one wants to use a returned action in an Effect cleanup / forked computation](https://github.com/thomashoneyman/purescript-halogen-hooks/issues/30#issuecomment-627712157)

Reply:
It's not necessary if you're modifying state (`modifyState` always uses fresh state), but it would be if you are relying on the `state` variable to say perform an effect in one of these actions. But then you have to basically guess whether it would ever make sense to do a particular action in an effect cleanup / forked computation. If the answer is yes, and you do rely on the `state` variable to perform an effect, then you should use `useGet`.

Reply:
> It's not necessary if you're modifying state (`modifyState` always uses fresh state), but it would be if you are relying on the state variable to say perform an effect in one of these actions.

Another situation that can arise is if you need to refer to the state that a library returns. For example, if I had a hook like this:
```purescript
state' <- do
  state /\ stateId <- useState 0
  useTickEffect { } $ Hooks.modify_ stateId (_ + 1)
  pure state
useTickEffect { } do
  pure $ Just do
    liftEffect $ log $ "This value is stale: " <> show state
```
... the state value is stale. There are two ways this could be fixed
1. the library **user** uses `getState <- useGet state`

```purescript
state <- do
  state /\ stateId <- useState 0
  useTickEffect { } $ Hooks.modify_ stateId (_ + 1)
  pure state
getState <- useGet state
useTickEffect { } do
  pure $ Just do
    state' <- getState
    liftEffect $ log $ "This value is stale: " <> show state
    liftEffect $ log $ "This value is not stale: <> show state'
```

2. the library **author** returns `{ state, getState: Hooks.get stateId }`
```purescript
{ state, getState } <- do
  state /\ stateId <- useState 0
  useTickEffect { } $ Hooks.modify_ stateId (_ + 1)
  pure { state, getState: Hooks.get stateId }
useTickEffect { } do
  pure $ Just do
    state' <- getState
    liftEffect $ log $ "This value is stale: " <> show state
    liftEffect $ log $ "This value is not stale: <> show state'
```

Reply:
With Hooks 0.4.0 this is back to the usual.

Reply:
I'll update this PR to use that version later on today.